---
title: "cellbaseR"
author: "Mohammed Omar Elsiddieg Abdallah"
date: "19 April 2016"
output: 
   rmarkdown::html_vignette
#  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{"cellbaseR"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Loading the library

```{r, eval=FALSE }
library(cellbaseR)

```

## The CellbaseQuery class

Defining the CellbaseQuery object holding the config to connect to cellbase data

```{r , eval=FALSE, message=FALSE}
# to get the default CellbaseQuery object (human data, from genome GRCh37)
library(cellbaseR)
cb <- CellbaseQuery()
# to change the default species from human to mouse for example
mm <-CellbaseQuery(species = "mmsculus") 

```

You can see the avaiable species and corresonding data for each species by
running the cbSpecies method passing  the CellbaseQuery object you just created 

```{r, eval=FALSE, message=FALSE, warning=FALSE}
library(cellbaseR)
cb <- CellbaseQuery()
sp <- cbSpecies(object = cb)
# This will give you a CellbaseResult object
# to get the dataframe of all available species
sp <- cellbaseData(sp)
names(sp)
# see all supported species
sp$scientificName
# See what categories of data are avaible for humans
sp$data[[1]]


```

## The CellbaseQuery Methods
In addition to the cbSpecies you have seen already, The cellbaseR package
provide many methods to query the cellbase webservices, they include:  
 
-  cbGene  
-  cbSnp  
-  cbProtein  
-  cbTranscript  
-  cbRegion  
-  cbVariant  
-  cbXref  
-  cbClinical  

In all cases the user is expected to provide the ids for the query and the 
resource to be queried, in addition to the CellbaseQuery object they created.
For example, a query through the cbGene will look something like this  
**cbGene**  

```{r, eval=FALSE}
cb <- CellbaseQuery()
genes <- c("TP73","TET1")

res <- cbGene(object = cb, ids = genes, resource = "transcript")
# to get the resulting data.frame run cellbaseData()
res <- cellbaseData(object = res)
names(res)
# as you can see the res dataframe also contains an exons column 
# which is in fact a list column of nested dataframes, to get the
# exons data.frame for first transcript
TET1_transcripts <- res$exons[[1]]
names(TET1_transcripts)
```


**cbRegion**  
```{r, eval=FALSE}
# making a query through cbRegion to get all the clinically relevant variants 
# in a specific region
res <- cbRegion(object=cb,ids="17:1000000-1100000",
resource="clinical")
# to get the data 
res <- cellbaseData(res)
# to get all conservation data in this region
res <- cbRegion(object=cb,ids="17:1000000-1100000",
resource="conservation")
#likewise to get all the regulatory data for the same region
res <- cbRegion(object=cb,ids="17:1000000-1100000",
resource="regulatory")

```
**cbVraiant**  
Getting annotation for a specific variant
```{r,eval=FALSE}
res2 <- cbVariant(object =cb,ids = "1:169549811:A:G",resource = "annotation")
# to get the data 
res2 <- cellbaseR::cellbaseData(res2)
```

A very powerfull feature of cellbaseR is ability to fetch a wealth of clinical 
data, as well as abiliy to fiter these data by gene, phenotype, rs, etcâ€¦

**cbClinical**

```{r, eval=FALSE}
# First we have to specify aour filters, we do that by creating an object of
# class CellbaseParam
cbparam <- CellbaseParam(gene=c("TP73","TET1"), genome="GRCh38")
cbparam
# Note that cbClinical does NOT require any Ids to be passed, only the filters
# and of course the CellbaseQuery object
res <- cbClinical(object=cb,filters=cbparam)

```


**annotateVCf**
```{r, eval=FALSE}


```


